litex/soc/integration/export.py

```diff
@@ -176,14 +191,8 @@ def get_csr_header(regions, constants, with_access_functions=True):
     if with_access_functions:
         r += "#include <stdint.h>\n"
         r += "#ifdef CSR_ACCESSORS_DEFINED\n"
-        r += "extern void csr_wr_uint8(uint8_t v, unsigned long a);\n"
-        r += "extern void csr_wr_uint16(uint16_t v, unsigned long a);\n"
-        r += "extern void csr_wr_uint32(uint32_t v, unsigned long a);\n"
-        r += "extern void csr_wr_uint64(uint64_t v, unsigned long a);\n"
-        r += "extern uint8_t csr_rd_uint8(unsigned long a);\n"
-        r += "extern uint16_t csr_rd_uint16(unsigned long a);\n"
-        r += "extern uint32_t csr_rd_uint32(unsigned long a);\n"
-        r += "extern uint64_t csr_rd_uint64(unsigned long a);\n"
+        r += "extern void csr_write_sub(unsigned long v, unsigned long a);\n"
+        r += "extern unsigned long csr_read_sub(unsigned long a);\n"
```

Comment:
These being marked as `extern` will mean the compiler can't optimize them without lto turned on (see https://github.com/enjoy-digital/litex/pull/253 for that).

Reply:
this is the case where they're externally provided by downstream, and downstream defines CSR_ACCESSORS_DEFINED to let us know we shouldn't be using the built-in ones.

That's the way it was before I ever got involved :) The litex-provided accessors (in common.h) are definitely _NOT_ marked as external